File: examples/23-udp-and-peer-discovery/go/main.go
// chatterd v2 (Go) — UDP multicast peer discovery + a TCP chat exchange.
// Grows the same chatterd program introduced in chapter 21: the TCP frame is
// the canonical chatterd chat frame (2-byte magic "CH" + 1-byte version +
// 1-byte type + 2-byte big-endian length + UTF-8 payload) shared by every
// chatterd version (ch21-ch27). The UDP discovery beacon below is a
// completely separate wire object — a plain ASCII datagram, no relation to
// the chat frame's magic/version/type/length header.
//
//	chatterd discover --group 239.7.7.7 --port 51888 --name alice \
//	    --tcp-port 9101 --iface 127.0.0.1 [--announce-ms 200] [--rounds 10]
//
// The multicast listener uses net.ListenMulticastUDP; the beacon sender is a
// raw socket whose IP_MULTICAST_IF/LOOP/TTL are set through golang.org/x/sys/
// unix. The accept and receive loops run as goroutines coordinated by a context
// and a WaitGroup; errors are wrapped with %w.
package main

import (
	"context"
	"fmt"
	"io"
	"net"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"golang.org/x/sys/unix"
)

const beaconMagic = "CHATTERD1" // UDP beacon tag only

// --- canonical chatterd chat frame (TCP) ------------------------------------
// magic 'C' 'H', version 0x01, type (1 byte), length (2 bytes big-endian,
// 0..65535), payload (UTF-8, type-specific). This header is byte-identical
// across every chatterd version; ch23's post-discovery greeting uses only
// JOIN (announce the dialer's name) and DELIVER (name + NUL + text, the
// listener's reply) — the same types chapters 21/22/24/27 share.
const (
	frameMagic0  = 0x43 // 'C'
	frameMagic1  = 0x48 // 'H'
	frameVersion = 0x01
)

const (
	frameJoin    = 1
	frameMsg     = 2
	frameDeliver = 3
	frameWelcome = 4
	framePing    = 5
)

const maxFramePayload = 0xFFFF

type config struct {
	group      string
	port       int
	name       string
	tcpPort    int
	iface      string
	announceMs int
	rounds     int
}

func usage() {
	fmt.Fprintln(os.Stderr,
		"usage: chatterd discover --group <ip> --port <n> --name <s> "+
			"[--tcp-port <n>] [--iface <ip>] [--announce-ms <n>] [--rounds <n>]")
}

// parseArgs returns (cfg, exitCode, ok). On any parse error it prints usage and
// returns ok=false with exit code 2.
func parseArgs(argv []string) (config, int, bool) {
	c := config{tcpPort: 9101, iface: "127.0.0.1", announceMs: 200, rounds: 10}
	if len(argv) < 2 || argv[1] != "discover" {
		usage()
		return c, 2, false
	}
	args := argv[2:]
	next := func(i *int, name string) (string, bool) {
		if *i+1 >= len(args) {
			fmt.Fprintf(os.Stderr, "chatterd: %s needs a value\n", name)
			return "", false
		}
		*i++
		return args[*i], true
	}
	posInt := func(s string, allowZero bool) (int, bool) {
		v, err := strconv.Atoi(s)
		if err != nil || v < 0 || (!allowZero && v == 0) {
			return 0, false
		}
		return v, true
	}
	port16 := func(s string) (int, bool) {
		v, err := strconv.Atoi(s)
		if err != nil || v < 1 || v > 65535 {
			return 0, false
		}
		return v, true
	}
	for i := 0; i < len(args); i++ {
		a := args[i]
		switch a {
		case "--group":
			v, ok := next(&i, a)
			if !ok {
				return c, 2, false
			}
			if ip := net.ParseIP(v); ip == nil || ip.To4() == nil {
				usage()
				return c, 2, false
			}
			c.group = v
		case "--name":
			v, ok := next(&i, a)
			if !ok {
				return c, 2, false
			}
			c.name = v
		case "--iface":
			v, ok := next(&i, a)
			if !ok {
				return c, 2, false
			}
			if ip := net.ParseIP(v); ip == nil || ip.To4() == nil {
				usage()
				return c, 2, false
			}
			c.iface = v
		case "--port":
			v, ok := next(&i, a)
			p, pok := port16(v)
			if !ok || !pok {
				usage()
				return c, 2, false
			}
			c.port = p
		case "--tcp-port":
			v, ok := next(&i, a)
			p, pok := port16(v)
			if !ok || !pok {
				usage()
				return c, 2, false
			}
			c.tcpPort = p
		case "--announce-ms":
			v, ok := next(&i, a)
			p, pok := posInt(v, false)
			if !ok || !pok {
				usage()
				return c, 2, false
			}
			c.announceMs = p
		case "--rounds":
			v, ok := next(&i, a)
			p, pok := posInt(v, false)
			if !ok || !pok {
				usage()
				return c, 2, false
			}
			c.rounds = p
		default:
			fmt.Fprintf(os.Stderr, "chatterd: unknown argument: %s\n", a)
			usage()
			return c, 2, false
		}
	}
	if c.group == "" || c.port == 0 || c.name == "" {
		usage()
		return c, 2, false
	}
	return c, 0, true
}

// --- canonical chatterd chat frame I/O ---------------------------------------

func writeFrame(w io.Writer, typ byte, payload []byte) error {
	if len(payload) > maxFramePayload {
		return fmt.Errorf("frame payload too large: %d", len(payload))
	}
	hdr := [6]byte{
		frameMagic0, frameMagic1, frameVersion, typ,
		byte(len(payload) >> 8), byte(len(payload)),
	}
	if _, err := w.Write(hdr[:]); err != nil {
		return fmt.Errorf("write frame header: %w", err)
	}
	if _, err := w.Write(payload); err != nil {
		return fmt.Errorf("write frame body: %w", err)
	}
	return nil
}

func readFrame(r io.Reader) (byte, []byte, error) {
	var hdr [6]byte
	if _, err := io.ReadFull(r, hdr[:]); err != nil {
		return 0, nil, fmt.Errorf("read frame header: %w", err)
	}
	if hdr[0] != frameMagic0 || hdr[1] != frameMagic1 {
		return 0, nil, fmt.Errorf("bad frame magic")
	}
	if hdr[2] != frameVersion {
		return 0, nil, fmt.Errorf("unsupported frame version: %d", hdr[2])
	}
	typ := hdr[3]
	n := int(hdr[4])<<8 | int(hdr[5])
	body := make([]byte, n)
	if _, err := io.ReadFull(r, body); err != nil {
		return 0, nil, fmt.Errorf("read frame body: %w", err)
	}
	return typ, body, nil
}

func interfaceForIP(ip net.IP) (*net.Interface, error) {
	ifaces, err := net.Interfaces()
	if err != nil {
		return nil, fmt.Errorf("list interfaces: %w", err)
	}
	for i := range ifaces {
		addrs, err := ifaces[i].Addrs()
		if err != nil {
			continue
		}
		for _, a := range addrs {
			var aip net.IP
			switch v := a.(type) {
			case *net.IPNet:
				aip = v.IP
			case *net.IPAddr:
				aip = v.IP
			}
			if aip.Equal(ip) {
				return &ifaces[i], nil
			}
		}
	}
	return nil, fmt.Errorf("no interface has address %s", ip)
}

// makeSender builds the raw UDP socket used to emit beacons, pinning the
// outgoing multicast interface and enabling local loopback so a second peer on
// the same host still receives our beacons.
func makeSender(ip4 [4]byte) (int, error) {
	fd, err := unix.Socket(unix.AF_INET, unix.SOCK_DGRAM, 0)
	if err != nil {
		return -1, fmt.Errorf("socket: %w", err)
	}
	if err := unix.SetsockoptInet4Addr(fd, unix.IPPROTO_IP, unix.IP_MULTICAST_IF, ip4); err != nil {
		unix.Close(fd)
		return -1, fmt.Errorf("IP_MULTICAST_IF: %w", err)
	}
	if err := unix.SetsockoptByte(fd, unix.IPPROTO_IP, unix.IP_MULTICAST_LOOP, 1); err != nil {
		unix.Close(fd)
		return -1, fmt.Errorf("IP_MULTICAST_LOOP: %w", err)
	}
	if err := unix.SetsockoptByte(fd, unix.IPPROTO_IP, unix.IP_MULTICAST_TTL, 1); err != nil {
		unix.Close(fd)
		return -1, fmt.Errorf("IP_MULTICAST_TTL: %w", err)
	}
	return fd, nil
}

func dial(ip string, port int) (net.Conn, error) {
	addr := net.JoinHostPort(ip, strconv.Itoa(port))
	// Retry: the peer's beacon may arrive a hair before its listener is ready.
	for i := 0; i < 40; i++ {
		c, err := net.DialTimeout("tcp", addr, 500*time.Millisecond)
		if err == nil {
			return c, nil
		}
		time.Sleep(50 * time.Millisecond)
	}
	return nil, fmt.Errorf("connect %s: giving up", addr)
}

func acceptLoop(ctx context.Context, wg *sync.WaitGroup, ln *net.TCPListener, deliverPayload []byte) {
	defer wg.Done()
	for {
		_ = ln.SetDeadline(time.Now().Add(200 * time.Millisecond))
		conn, err := ln.Accept()
		if err != nil {
			if ctx.Err() != nil {
				return
			}
			continue
		}
		// Read the dialer's JOIN frame, reply with our DELIVER frame (our
		// name + NUL + greeting text), close.
		if _, _, rerr := readFrame(conn); rerr == nil {
			_ = writeFrame(conn, frameDeliver, deliverPayload)
		}
		conn.Close()
	}
}

func recvLoop(ctx context.Context, wg *sync.WaitGroup, udp *net.UDPConn, cfg config) {
	defer wg.Done()
	seen := make(map[string]bool)
	buf := make([]byte, 2048)
	for {
		_ = udp.SetReadDeadline(time.Now().Add(200 * time.Millisecond))
		n, _, err := udp.ReadFromUDP(buf)
		if err != nil {
			if ctx.Err() != nil {
				return
			}
			continue
		}
		tok := strings.Fields(string(buf[:n]))
		if len(tok) != 4 || tok[0] != beaconMagic {
			continue
		}
		pname := tok[1]
		if pname == cfg.name || seen[pname] { // our own beacon or already known
			continue
		}
		pport, err := strconv.Atoi(tok[2])
		if err != nil || pport < 1 || pport > 65535 {
			continue
		}
		pip := tok[3]
		seen[pname] = true

		fmt.Printf("discovered peer %s at %s:%d\n", pname, pip, pport)
		conn, err := dial(pip, pport)
		if err != nil {
			fmt.Fprintf(os.Stderr, "chatterd: error: dial %s: %v\n", pname, err)
			continue
		}
		// JOIN (announce ourselves), then read the peer's DELIVER reply.
		if err := writeFrame(conn, frameJoin, []byte(cfg.name)); err != nil {
			fmt.Fprintf(os.Stderr, "chatterd: error: %v\n", err)
			conn.Close()
			continue
		}
		_, payload, err := readFrame(conn)
		conn.Close()
		if err != nil {
			fmt.Fprintf(os.Stderr, "chatterd: error: %v\n", err)
			continue
		}
		parts := strings.SplitN(string(payload), "\x00", 2)
		rname, rtext := parts[0], ""
		if len(parts) == 2 {
			rtext = parts[1]
		}
		fmt.Printf("peer %s says: %s\n", rname, rtext)
	}
}

func discover(cfg config) error {
	ip := net.ParseIP(cfg.iface)
	if ip == nil || ip.To4() == nil {
		return fmt.Errorf("bad --iface address: %s", cfg.iface)
	}
	gip := net.ParseIP(cfg.group)
	if gip == nil || gip.To4() == nil {
		return fmt.Errorf("bad --group address: %s", cfg.group)
	}
	var ip4, group4 [4]byte
	copy(ip4[:], ip.To4())
	copy(group4[:], gip.To4())

	ifi, err := interfaceForIP(ip)
	if err != nil {
		return err
	}

	lnAny, err := net.Listen("tcp", net.JoinHostPort(cfg.iface, strconv.Itoa(cfg.tcpPort)))
	if err != nil {
		return fmt.Errorf("listen tcp: %w", err)
	}
	ln := lnAny.(*net.TCPListener)
	defer ln.Close()

	udp, err := net.ListenMulticastUDP("udp4", ifi, &net.UDPAddr{IP: gip, Port: cfg.port})
	if err != nil {
		return fmt.Errorf("listen multicast: %w", err)
	}
	defer udp.Close()

	sfd, err := makeSender(ip4)
	if err != nil {
		return err
	}
	defer unix.Close(sfd)

	fmt.Fprintf(os.Stderr, "chatterd: announcing as %s on %s:%d (tcp %s:%d)\n",
		cfg.name, cfg.group, cfg.port, cfg.iface, cfg.tcpPort)

	deliverPayload := []byte(cfg.name + "\x00hello from " + cfg.name)
	ctx, cancel := context.WithCancel(context.Background())
	var wg sync.WaitGroup
	wg.Add(2)
	go acceptLoop(ctx, &wg, ln, deliverPayload)
	go recvLoop(ctx, &wg, udp, cfg)

	beacon := []byte(fmt.Sprintf("%s %s %d %s", beaconMagic, cfg.name, cfg.tcpPort, cfg.iface))
	dst := &unix.SockaddrInet4{Port: cfg.port, Addr: group4}
	for r := 0; r < cfg.rounds; r++ {
		if err := unix.Sendto(sfd, beacon, 0, dst); err != nil {
			cancel()
			wg.Wait()
			return fmt.Errorf("sendto beacon: %w", err)
		}
		time.Sleep(time.Duration(cfg.announceMs) * time.Millisecond)
	}
	// Grace window so final exchanges complete, then unblock the loops.
	time.Sleep(800 * time.Millisecond)
	cancel()
	ln.Close()
	udp.Close()
	wg.Wait()
	return nil
}

func main() {
	cfg, code, ok := parseArgs(os.Args)
	if !ok {
		os.Exit(code)
	}
	if err := discover(cfg); err != nil {
		fmt.Fprintf(os.Stderr, "chatterd: error: %v\n", err)
		os.Exit(1)
	}
}

File: examples/23-udp-and-peer-discovery/rust/src/main.rs
// chatterd v2 (Rust 2024) — UDP multicast peer discovery + a TCP chat
// exchange. Grows the same chatterd program introduced in chapter 21: the TCP
// frame is the canonical chatterd chat frame (2-byte magic "CH" + 1-byte
// version + 1-byte type + 2-byte big-endian length + UTF-8 payload) shared by
// every chatterd version (ch21-ch27). The UDP discovery beacon below is a
// completely separate wire object — a plain ASCII datagram, no relation to
// the chat frame's magic/version/type/length header.
//
//   chatterd discover --group 239.7.7.7 --port 51888 --name alice \
//       --tcp-port 9101 --iface 127.0.0.1 [--announce-ms 200] [--rounds 10]
//
// The multicast socket is built with the socket2 crate (IP_ADD_MEMBERSHIP /
// IP_MULTICAST_IF / IP_MULTICAST_LOOP), then converted into a std UdpSocket
// (an OwnedFd) shared across threads. Fallible paths use Result/?; the accept
// and receive loops run on std threads coordinated by an AtomicBool.

use std::collections::HashSet;
use std::error::Error;
use std::io::{self, Read, Write};
use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4, TcpListener, TcpStream, UdpSocket};
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, Ordering};
use std::thread;
use std::time::Duration;

use socket2::{Domain, Protocol, SockAddr, Socket, Type};

const BEACON_MAGIC: &str = "CHATTERD1"; // UDP beacon tag only

// --- canonical chatterd chat frame (TCP) -----------------------------------
// magic 'C' 'H', version 0x01, type (1 byte), length (2 bytes big-endian,
// 0..65535), payload (UTF-8, type-specific). This header is byte-identical
// across every chatterd version; ch23's post-discovery greeting uses only
// JOIN (announce the dialer's name) and DELIVER (name + NUL + text, the
// listener's reply) — the same types chapters 21/22/24/27 share.
const FRAME_MAGIC: [u8; 2] = [0x43, 0x48]; // "CH"
const FRAME_VERSION: u8 = 0x01;
const FRAME_JOIN: u8 = 1;
#[allow(dead_code)]
const FRAME_MSG: u8 = 2;
const FRAME_DELIVER: u8 = 3;
#[allow(dead_code)]
const FRAME_WELCOME: u8 = 4;
#[allow(dead_code)]
const FRAME_PING: u8 = 5;
const MAX_FRAME_PAYLOAD: usize = 0xFFFF;

type Fallible<T> = Result<T, Box<dyn Error + Send + Sync>>;

#[derive(Clone)]
struct Config {
    group: Ipv4Addr,
    port: u16,
    name: String,
    tcp_port: u16,
    iface: Ipv4Addr,
    announce_ms: u64,
    rounds: u32,
}

fn usage() {
    eprintln!(
        "usage: chatterd discover --group <ip> --port <n> --name <s> \
[--tcp-port <n>] [--iface <ip>] [--announce-ms <n>] [--rounds <n>]"
    );
}

fn parse_port(s: &str) -> Result<u16, i32> {
    s.parse::<u16>().ok().filter(|&n| n >= 1).ok_or_else(|| {
        usage();
        2
    })
}

/// Parse argv. On any error prints usage and returns Err(exit_code).
fn parse_args() -> Result<Config, i32> {
    let argv: Vec<String> = std::env::args().collect();
    if argv.len() < 2 || argv[1] != "discover" {
        usage();
        return Err(2);
    }
    let mut group: Option<Ipv4Addr> = None;
    let mut port: Option<u16> = None;
    let mut name: Option<String> = None;
    let mut tcp_port: u16 = 9101;
    let mut iface: Ipv4Addr = Ipv4Addr::LOCALHOST;
    let mut announce_ms: u64 = 200;
    let mut rounds: u32 = 10;

    let args = &argv[2..];
    let take = |i: &mut usize, flag: &str| -> Result<String, i32> {
        if *i + 1 >= args.len() {
            eprintln!("chatterd: {flag} needs a value");
            return Err(2);
        }
        *i += 1;
        Ok(args[*i].clone())
    };
    let mut i = 0;
    while i < args.len() {
        let a = args[i].as_str();
        match a {
            "--group" => {
                let v = take(&mut i, a)?;
                group = Some(v.parse().map_err(|_| {
                    usage();
                    2
                })?);
            }
            "--name" => name = Some(take(&mut i, a)?),
            "--iface" => {
                let v = take(&mut i, a)?;
                iface = v.parse().map_err(|_| {
                    usage();
                    2
                })?;
            }
            "--port" => {
                let v = take(&mut i, a)?;
                port = Some(parse_port(&v)?);
            }
            "--tcp-port" => {
                let v = take(&mut i, a)?;
                tcp_port = parse_port(&v)?;
            }
            "--announce-ms" => {
                let v = take(&mut i, a)?;
                announce_ms = v.parse::<u64>().ok().filter(|&n| n > 0).ok_or_else(|| {
                    usage();
                    2
                })?;
            }
            "--rounds" => {
                let v = take(&mut i, a)?;
                rounds = v.parse::<u32>().ok().filter(|&n| n > 0).ok_or_else(|| {
                    usage();
                    2
                })?;
            }
            _ => {
                eprintln!("chatterd: unknown argument: {a}");
                usage();
                return Err(2);
            }
        }
        i += 1;
    }

    match (group, port, name) {
        (Some(group), Some(port), Some(name)) => Ok(Config {
            group,
            port,
            name,
            tcp_port,
            iface,
            announce_ms,
            rounds,
        }),
        _ => {
            usage();
            Err(2)
        }
    }
}

// --- canonical chatterd chat frame I/O --------------------------------------

fn write_frame(w: &mut impl Write, typ: u8, payload: &[u8]) -> io::Result<()> {
    if payload.len() > MAX_FRAME_PAYLOAD {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("frame payload too large: {}", payload.len()),
        ));
    }
    let mut hdr = [0u8; 6];
    hdr[0] = FRAME_MAGIC[0];
    hdr[1] = FRAME_MAGIC[1];
    hdr[2] = FRAME_VERSION;
    hdr[3] = typ;
    hdr[4] = (payload.len() >> 8) as u8;
    hdr[5] = (payload.len() & 0xFF) as u8;
    w.write_all(&hdr)?;
    w.write_all(payload)?;
    Ok(())
}

fn read_frame(r: &mut impl Read) -> io::Result<(u8, Vec<u8>)> {
    let mut hdr = [0u8; 6];
    r.read_exact(&mut hdr)?;
    if hdr[0] != FRAME_MAGIC[0] || hdr[1] != FRAME_MAGIC[1] {
        return Err(io::Error::new(io::ErrorKind::InvalidData, "bad frame magic"));
    }
    if hdr[2] != FRAME_VERSION {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            format!("unsupported frame version: {}", hdr[2]),
        ));
    }
    let typ = hdr[3];
    let n = ((hdr[4] as usize) << 8) | hdr[5] as usize;
    let mut body = vec![0u8; n];
    r.read_exact(&mut body)?;
    Ok((typ, body))
}

/// Build the multicast UDP socket and hand back a std UdpSocket (an OwnedFd).
fn make_multicast(cfg: &Config) -> Fallible<UdpSocket> {
    let sock = Socket::new(Domain::IPV4, Type::DGRAM, Some(Protocol::UDP))?;
    sock.set_reuse_address(true)?;
    let bind = SocketAddrV4::new(Ipv4Addr::UNSPECIFIED, cfg.port);
    sock.bind(&SockAddr::from(bind))?;
    sock.join_multicast_v4(&cfg.group, &cfg.iface)?;
    sock.set_multicast_if_v4(&cfg.iface)?;
    sock.set_multicast_loop_v4(true)?;
    sock.set_multicast_ttl_v4(1)?;
    sock.set_read_timeout(Some(Duration::from_millis(200)))?;
    Ok(sock.into())
}

fn dial(ip: Ipv4Addr, port: u16) -> Fallible<TcpStream> {
    let addr = SocketAddr::from((ip, port));
    // Retry: the peer's beacon may arrive a hair before its listener is ready.
    for _ in 0..40 {
        if let Ok(s) = TcpStream::connect_timeout(&addr, Duration::from_millis(500)) {
            return Ok(s);
        }
        thread::sleep(Duration::from_millis(50));
    }
    Err(format!("connect {addr}: giving up").into())
}

fn accept_loop(listener: TcpListener, deliver_payload: Vec<u8>, stop: Arc<AtomicBool>) {
    listener.set_nonblocking(true).ok();
    while !stop.load(Ordering::Relaxed) {
        match listener.accept() {
            Ok((mut conn, _)) => {
                // Read the dialer's JOIN frame, reply with our DELIVER frame
                // (our name + NUL + greeting text), close.
                if read_frame(&mut conn).is_ok() {
                    let _ = write_frame(&mut conn, FRAME_DELIVER, &deliver_payload);
                }
            }
            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
                thread::sleep(Duration::from_millis(50));
            }
            Err(_) => thread::sleep(Duration::from_millis(50)),
        }
    }
}

fn recv_loop(udp: Arc<UdpSocket>, cfg: Config, stop: Arc<AtomicBool>) {
    let mut seen: HashSet<String> = HashSet::new();
    let mut buf = [0u8; 2048];
    while !stop.load(Ordering::Relaxed) {
        let n = match udp.recv_from(&mut buf) {
            Ok((n, _)) => n,
            Err(ref e)
                if e.kind() == io::ErrorKind::WouldBlock
                    || e.kind() == io::ErrorKind::TimedOut =>
            {
                continue;
            }
            Err(_) => continue,
        };
        let msg = String::from_utf8_lossy(&buf[..n]);
        let tok: Vec<&str> = msg.split_whitespace().collect();
        if tok.len() != 4 || tok[0] != BEACON_MAGIC {
            continue;
        }
        let pname = tok[1];
        if pname == cfg.name || seen.contains(pname) {
            continue; // our own beacon or already known
        }
        let (pport, pip) = match (tok[2].parse::<u16>(), tok[3].parse::<Ipv4Addr>()) {
            (Ok(p), Ok(ip)) if p >= 1 => (p, ip),
            _ => continue,
        };
        seen.insert(pname.to_string());

        println!("discovered peer {pname} at {pip}:{pport}");
        let mut conn = match dial(pip, pport) {
            Ok(c) => c,
            Err(e) => {
                eprintln!("chatterd: error: dial {pname}: {e}");
                continue;
            }
        };
        // JOIN (announce ourselves), then read the peer's DELIVER reply.
        if let Err(e) = write_frame(&mut conn, FRAME_JOIN, cfg.name.as_bytes()) {
            eprintln!("chatterd: error: {e}");
            continue;
        }
        let (_typ, payload) = match read_frame(&mut conn) {
            Ok(r) => r,
            Err(e) => {
                eprintln!("chatterd: error: {e}");
                continue;
            }
        };
        let reply = String::from_utf8_lossy(&payload);
        let (rname, rtext) = reply.split_once('\0').unwrap_or((reply.as_ref(), ""));
        println!("peer {rname} says: {rtext}");
    }
}

fn discover(cfg: &Config) -> Fallible<()> {
    let listener = TcpListener::bind((cfg.iface, cfg.tcp_port))
        .map_err(|e| format!("listen tcp {}:{}: {e}", cfg.iface, cfg.tcp_port))?;
    let udp = Arc::new(make_multicast(cfg)?);

    eprintln!(
        "chatterd: announcing as {} on {}:{} (tcp {}:{})",
        cfg.name, cfg.group, cfg.port, cfg.iface, cfg.tcp_port
    );

    let deliver_payload = format!("{}\0hello from {}", cfg.name, cfg.name).into_bytes();
    let stop = Arc::new(AtomicBool::new(false));

    let acc = {
        let (p, s) = (deliver_payload.clone(), stop.clone());
        thread::spawn(move || accept_loop(listener, p, s))
    };
    let rcv = {
        let (u, c, s) = (udp.clone(), cfg.clone(), stop.clone());
        thread::spawn(move || recv_loop(u, c, s))
    };

    let beacon = format!("{} {} {} {}", BEACON_MAGIC, cfg.name, cfg.tcp_port, cfg.iface);
    let dst = SocketAddr::from((cfg.group, cfg.port));
    for _ in 0..cfg.rounds {
        udp.send_to(beacon.as_bytes(), dst)?;
        thread::sleep(Duration::from_millis(cfg.announce_ms));
    }
    // Grace window so final exchanges complete, then stop the loops.
    thread::sleep(Duration::from_millis(800));
    stop.store(true, Ordering::Relaxed);
    let _ = acc.join();
    let _ = rcv.join();
    Ok(())
}

fn main() {
    let code = match parse_args() {
        Ok(cfg) => match discover(&cfg) {
            Ok(()) => 0,
            Err(e) => {
                eprintln!("chatterd: error: {e}");
                1
            }
        },
        Err(code) => code,
    };
    std::process::exit(code);
}
